pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -0,0 +1,319 @@
+		"reverse_OctetDeltaCount",
+	}
+	AntreaInfoElements = []string{
+		"sourcePodName",
+		"sourcePodNamespace",
+		"sourceNodeName",
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+	}
+)
+
+var _ FlowExporter = new(flowExporter)
+
+type FlowExporter interface {
+	Run(stopCh <-chan struct{}, pollDone <-chan bool)
+}
+
+type flowExporter struct {
+	flowRecords    flowrecords.FlowRecords
+	process        ipfix.IPFIXExportingProcess
+	elementsList   []*ipfixentities.InfoElement
+	exportInterval time.Duration
+	templateID     uint16
+}
+
+func genObservationID() (uint32, error) {
+	name, err := env.GetNodeName()
+	if err != nil {
+		return 0, err
+	}
+	h := fnv.New32()
+	h.Write([]byte(name))
+	return h.Sum32(), nil
+}
+
+func InitFlowExporter(collector net.Addr, records flowrecords.FlowRecords, expInterval time.Duration) (*flowExporter, error) {
+	// Create IPFIX exporting expProcess and initialize registries and other related entities
+	obsID, err := genObservationID()
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate obsID for IPFIX ipfixexport: %v", err)
+	}
+
+	var expProcess ipfix.IPFIXExportingProcess
+	if collector.Network() == "tcp" {
+		// TCP transport do not need any tempRefTimeout, so sending 0.
+		expProcess, err = ipfix.NewIPFIXExportingProcess(collector, obsID, 0)
+	} else {
+		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
+		expProcess, err = ipfix.NewIPFIXExportingProcess(collector, obsID, 1800)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error while initializing IPFIX exporting expProcess: %v", err)
+	}
+	expProcess.LoadRegistries()
+
+	flowExp := &flowExporter{
+		records,
+		expProcess,
+		nil,
+		expInterval,
+		0,
+	}
+
+	flowExp.templateID = flowExp.process.NewTemplateID()
+	templateRec := ipfix.NewIPFIXTemplateRecord(uint16(len(IANAInfoElements)+len(IANAReverseInfoElements)+len(AntreaInfoElements)), flowExp.templateID)
+
+	sentBytes, err := flowExp.sendTemplateRecord(templateRec)
+	if err != nil {
+		return nil, fmt.Errorf("error while creating and sending template record through IPFIX process: %v", err)
+	}
+	klog.V(2).Infof("Initialized flow exporter and sent %d bytes size of template record", sentBytes)
+
+	return flowExp, nil
+}
+
+func (exp *flowExporter) Run(stopCh <-chan struct{}, pollDone <-chan bool) {
+	klog.Infof("Start exporting IPFIX flow records")
+	ticker := time.NewTicker(exp.exportInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			exp.process.CloseConnToCollector()
+			break
+		case <-ticker.C:
+			// Waiting for pollDone channel is necessary because IPFIX collector computes throughput based on
+			// flow records received interval, therefore we need to wait for poll go routine (ConnectionStore.Run).
+			// pollDone provides this synchronization. Note that export interval is multiple of poll interval, and poll
+			// interval is at least one second long.
+			<-pollDone
+			err := exp.flowRecords.BuildFlowRecords()
```

Comment:
that's a bit hacky IMO. There is no synchronization guarantees. We call `BuildFlowRecords` (which calls `IterateCxnMapWithCB `) and in theory another polling cycle can begin while this function is called.

I think we should still have a mutex and proper synchronization so that the interface is easy to reason about and `IterateCxnMapWithCB` can be called by other components if needed (or more generally other components can have access to the connection map). A mutex / lock is basically "free" when there is no contention and it seems that with your scheme you drastically reduce the potential for contention. However, I don't think your scheme gives synchronization guarantees, unless I am missing something.

Reply:
Hi @antoninbas, Thanks for the comments.
Yes, synchronization between those poll and export go-routines is not complete. However, the poll interval is restricted to be at least one second long. I felt it is very unlikely that export go-routine is not scheduled and BuildFlowRecords, which is the first task of go-routine could not be done within one second. I agree, in theory, it is still possible that export (BuildFlowRecords) go-routine may not be done in time.

I actually thought of adding mutex for the connection map to avoid the race condition, where poll go-routine may start before the export routine is done. I also felt I may add confusion by using both channel and mutex for synchronization, as the one-second poll interval is large enough. I agree that the map can be accessed by some other components. I will add the mutex for the connection map. Hope this sounds good.

Reply:
I did a small experiment: Defined a variable called pollCycle in ConnectionStore data structure, incrementing this in poll goroutine after dumping/updating of flows in done and printing this in export goroutine, while protecting it with a lock.
For the poll interval of 1s and export interval of 5s: Without channel and just mutex, export goroutine exports stats/data from 4th pollCycle instead of 5th. To get the order right, channel is helpful.

Reply:
I didn't mind the channel, it just needed a mutex as well.